daemon/SecretsImpl/applicationpermissions.cpp

```diff
@@ -12,46 +12,125 @@
+                    return retn;
+                }
+            }
+        }
 
-    // special case handling of the pid 999999 for demonstration testing purposes only!
-    // TODO: remove this special case handling!
-    if (pid == 999999) {
-        qWarning() << "WARNING! Special case pid handling needs to be removed once access control / application management API exists!";
-        return QLatin1String("test-third-party-application");
+        return QString();
     }
 
+    QString readExeSymlink(pid_t pid)
+    {
+        struct stat sb;
+        const QString pidFile(QStringLiteral("/proc/%1/exe").arg(pid));
+        if (lstat(pidFile.toStdString().c_str(), &sb) == -1) {
+            qCDebug(lcSailfishSecretsDaemon) << "unable to lstat procfs exe link file for process:" << pid;
+            return QString();
+        }
+
+        QScopedPointer<char, QScopedPointerPodDeleter> linkName(reinterpret_cast<char*>(malloc(sb.st_size + 1)));
+        if (linkName.data() == NULL) {
+            qCDebug(lcSailfishSecretsDaemon) << "unable to allocate memory for link name for process:" << pid;
+            return QString();
+        }
+
+        ssize_t r = readlink(pidFile.toStdString().c_str(),
+                             linkName.data(),
+                             sb.st_size + 1);
+        if (r < 0) {
+            qCDebug(lcSailfishSecretsDaemon) << "unable to readlink exe for process:" << pid;
+            return QString();
+        }
+
+        linkName.data()[sb.st_size] = '\0';
+        const QString retn(QString::fromUtf8(QByteArray(linkName.data())));
+        qCDebug(lcSailfishSecretsDaemon) << "caller with pid" << pid << "has exe applicationId:" << retn;
+        return retn;
+    }
+
+    QString readCmdline(pid_t pid)
+    {
+        const QString pidFile(QStringLiteral("/proc/%1/cmdline").arg(pid));
+        if (!QFile::exists(pidFile)) {
+            qCDebug(lcSailfishSecretsDaemon) << "no cmdline in procfs for process:" << pid;
+            return QString();
+        }
+
+        QFile file(pidFile);
+        if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
+            qCDebug(lcSailfishSecretsDaemon) << "unable to open cmdline file:" << pidFile;
+            return QString();
+        }
+
+        QByteArray contents;
+        char data;
+        while (file.read(&data, 1)) {
+            if (data == '\0') {
+                data = ' ';
+            }
+            contents.append(data);
+        }
+
+        const QString retn(QString::fromUtf8(contents).trimmed());
+        qCDebug(lcSailfishSecretsDaemon) << "caller with pid" << pid << "has cmdline applicationId:" << retn;
+        return retn;
+    }
+}
+
+QString Sailfish::Secrets::Daemon::ApiImpl::ApplicationPermissions::applicationId(pid_t pid) const
+{
     if (pid == 0) {
         qCDebug(lcSailfishSecretsDaemon) << "zero pid, assuming privileged!";
```

Comment:
Is this still a correct assumption here?

Reply:
Yes, this should not be any different now.